src/uct/ugni/smsg/ugni_smsg_iface.c

```diff
@@ -265,19 +265,22 @@ static ucs_status_t uct_ugni_smsg_ep_flush(uct_ep_h tl_ep, unsigned flags,
                                            uct_completion_t *comp)
 {
     uct_ugni_smsg_ep_t *ep = ucs_derived_of(tl_ep, uct_ugni_smsg_ep_t);
+    ucs_status_t status;
 
-    if (comp != NULL) {
-        return UCS_ERR_UNSUPPORTED;
-    }
-
-    if (0 == ep->super.outstanding) {
+    if ((0 == ep->super.outstanding) && uct_ugni_can_send(&ep->super)) {
         /* We progress the local CQ anyways because we may get back send credits */
         progress_local_cq(ucs_derived_of(tl_ep->iface, uct_ugni_smsg_iface_t));
         UCT_TL_IFACE_STAT_FLUSH(ucs_derived_of(tl_ep->iface, uct_base_iface_t));
-        return UCS_OK;
+        status = UCS_OK;
     } else {
-        return uct_ugni_smsg_iface_flush(tl_ep->iface, 0, NULL);
+        if (comp != NULL){
```

Comment:
I thought that it had to return NO_RESOURCE so that UCP would put the completion in the pending queue? Without the check it wouldn't know to return NO_RESOURCE. But I can probably simplify the if structure by excising the progress_local_cq to the top.